File: rust-cli-version/src/models/orbit.rs
use serde::{Deserialize, Serialize};
use crate::models::World;

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum OrbitType {
    Empty,
    World,
    GasGiant,
    Belt,
    Hostile,
    Rockball,
}

pub trait Orbit {
    fn orbit_number(&self) -> u8;
    fn au(&self) -> f64;
    fn orbit_type(&self) -> OrbitType;
    fn to_ascii(&self) -> String;
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum OrbitContent {
    Empty(EmptyOrbit),
    World(WorldOrbit),
    GasGiant(GasGiant),
    Belt(Belt),
    Hostile(Hostile),
    Rockball(Rockball),
}

impl Orbit for OrbitContent {
    fn orbit_number(&self) -> u8 {
        match self {
            OrbitContent::Empty(o) => o.orbit_number(),
            OrbitContent::World(o) => o.orbit_number(),
            OrbitContent::GasGiant(o) => o.orbit_number(),
            OrbitContent::Belt(o) => o.orbit_number(),
            OrbitContent::Hostile(o) => o.orbit_number(),
            OrbitContent::Rockball(o) => o.orbit_number(),
        }
    }
    
    fn au(&self) -> f64 {
        match self {
            OrbitContent::Empty(o) => o.au(),
            OrbitContent::World(o) => o.au(),
            OrbitContent::GasGiant(o) => o.au(),
            OrbitContent::Belt(o) => o.au(),
            OrbitContent::Hostile(o) => o.au(),
            OrbitContent::Rockball(o) => o.au(),
        }
    }
    
    fn orbit_type(&self) -> OrbitType {
        match self {
            OrbitContent::Empty(o) => o.orbit_type(),
            OrbitContent::World(o) => o.orbit_type(),
            OrbitContent::GasGiant(o) => o.orbit_type(),
            OrbitContent::Belt(o) => o.orbit_type(),
            OrbitContent::Hostile(o) => o.orbit_type(),
            OrbitContent::Rockball(o) => o.orbit_type(),
        }
    }
    
    fn to_ascii(&self) -> String {
        match self {
            OrbitContent::Empty(o) => o.to_ascii(),
            OrbitContent::World(o) => o.to_ascii(),
            OrbitContent::GasGiant(o) => o.to_ascii(),
            OrbitContent::Belt(o) => o.to_ascii(),
            OrbitContent::Hostile(o) => o.to_ascii(),
            OrbitContent::Rockball(o) => o.to_ascii(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmptyOrbit {
    pub orbit_number: u8,
    pub au: f64,
}

impl Orbit for EmptyOrbit {
    fn orbit_number(&self) -> u8 { self.orbit_number }
    fn au(&self) -> f64 { self.au }
    fn orbit_type(&self) -> OrbitType { OrbitType::Empty }
    fn to_ascii(&self) -> String { "-".to_string() }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WorldOrbit {
    pub orbit_number: u8,
    pub au: f64,
    pub world: World,
}

impl Orbit for WorldOrbit {
    fn orbit_number(&self) -> u8 { self.orbit_number }
    fn au(&self) -> f64 { self.au }
    fn orbit_type(&self) -> OrbitType { OrbitType::World }
    fn to_ascii(&self) -> String { "W".to_string() }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GasGiant {
    pub orbit_number: u8,
    pub au: f64,
    pub size: GiantSize,
    pub moons: Vec<Moon>,
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub enum GiantSize {
    Small,
    Large,
}

impl Orbit for GasGiant {
    fn orbit_number(&self) -> u8 { self.orbit_number }
    fn au(&self) -> f64 { self.au }
    fn orbit_type(&self) -> OrbitType { OrbitType::GasGiant }
    fn to_ascii(&self) -> String {
        match self.size {
            GiantSize::Small => "G".to_string(),
            GiantSize::Large => "G".to_string(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Belt {
    pub orbit_number: u8,
    pub au: f64,
}

impl Orbit for Belt {
    fn orbit_number(&self) -> u8 { self.orbit_number }
    fn au(&self) -> f64 { self.au }
    fn orbit_type(&self) -> OrbitType { OrbitType::Belt }
    fn to_ascii(&self) -> String { "B".to_string() }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Hostile {
    pub orbit_number: u8,
    pub au: f64,
    pub atmosphere: u8,
    pub hydrographics: u8,
}

impl Orbit for Hostile {
    fn orbit_number(&self) -> u8 { self.orbit_number }
    fn au(&self) -> f64 { self.au }
    fn orbit_type(&self) -> OrbitType { OrbitType::Hostile }
    fn to_ascii(&self) -> String { "H".to_string() }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Rockball {
    pub orbit_number: u8,
    pub au: f64,
}

impl Orbit for Rockball {
    fn orbit_number(&self) -> u8 { self.orbit_number }
    fn au(&self) -> f64 { self.au }
    fn orbit_type(&self) -> OrbitType { OrbitType::Rockball }
    fn to_ascii(&self) -> String { "R".to_string() }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Moon {
    pub orbit: u8,
    pub size: u8,
}

File: go-version/pkg/models/orbit.go
package models

import (
	"fmt"
)

type OrbitType string

const (
	OrbitEmpty     OrbitType = "."
	OrbitBelt      OrbitType = "B"
	OrbitRockball  OrbitType = "R"
	OrbitHostile   OrbitType = "H"
	OrbitGasGiant  OrbitType = "G"
	OrbitWorld     OrbitType = "W"
	OrbitCompanion OrbitType = "S"
)

type Orbit interface {
	GetOrbitNumber() int
	GetAU() float64
	GetKid() OrbitType
	GetUWP() string
	IsWorld() bool
	ToASCII() string
}

type BaseOrbit struct {
	Star        *Star     `json:"-"`
	OrbitNumber int       `json:"orbit_number"`
	AU          float64   `json:"au"`
	Kid         OrbitType `json:"orbit_type"`
	Zone        int       `json:"zone"`
	Distant     bool      `json:"distant"`
	Port        string    `json:"starport"`
	Size        int       `json:"size"`
	Atmosphere  int       `json:"atmosphere"`
	Hydro       int       `json:"hydrographics"`
	Population  int       `json:"population"`
	Government  int       `json:"government"`
	Law         int       `json:"law_level"`
	Tech        int       `json:"tech_level"`
	Moons       []Moon    `json:"moons,omitempty"`
}

func (o *BaseOrbit) GetOrbitNumber() int { return o.OrbitNumber }
func (o *BaseOrbit) GetAU() float64      { return o.AU }
func (o *BaseOrbit) GetKid() OrbitType   { return o.Kid }
func (o *BaseOrbit) IsWorld() bool       { return o.Kid == OrbitWorld }

func (o *BaseOrbit) GetUWP() string {
	if o.Kid == "." {
		return ".......-."
	}
	return fmt.Sprintf("%s%s%s%s%s%s%s-%s",
		o.Port,
		toHex(o.Size),
		toHex(o.Atmosphere),
		toHex(o.Hydro),
		toHex(o.Population),
		toHex(o.Government),
		toHex(o.Law),
		toHex(o.Tech))
}

func (o *BaseOrbit) ToASCII() string {
	bio := " "
	if o.Zone == 0 {
		bio = "*"
	}
	if o.AU > o.Star.OuterLimit() {
		bio = "-"
	}
	output := fmt.Sprintf("  -- %2d. %s  %s // %s // %4.1f au",
		o.OrbitNumber+1, bio, o.Kid, o.GetUWP(), o.AU)
	
	for _, moon := range o.Moons {
		output += moon.ToASCII()
	}
	return output
}

func (o *BaseOrbit) IsInner() bool   { return o.Zone < 0 }
func (o *BaseOrbit) IsOuter() bool   { return o.Zone > 0 }
func (o *BaseOrbit) IsBiozone() bool { return o.Zone == 0 }
func (o *BaseOrbit) IsDistant() bool { return o.Distant }
func (o *BaseOrbit) IsLimit() bool   { return o.AU < o.Star.InnerLimit() }

type EmptyOrbit struct {
	BaseOrbit
}

type Belt struct {
	BaseOrbit
}

func (b *Belt) GetUWP() string {
	return "XR00000-0"
}

type Rockball struct {
	BaseOrbit
}

type Hostile struct {
	BaseOrbit
}

type GasGiant struct {
	BaseOrbit
	GiantSize string `json:"giant_size"`
}

func (g *GasGiant) GetUWP() string {
	if g.GiantSize == "S" {
		return "Small GG "
	}
	return "Large GG "
}

type World struct {
	BaseOrbit
	Temperature string   `json:"temperature"`
	Factions    []string `json:"factions,omitempty"`
	GasGiant    string   `json:"gas_giant,omitempty"`
	TradeCodes  []string `json:"trade_codes,omitempty"`
	Bases       string   `json:"bases,omitempty"`
	TravelCode  string   `json:"travel_code,omitempty"`
}

func (w *World) GetUWP() string {
	return fmt.Sprintf("%s%s%s%s%s%s%s-%s",
		w.Port,
		toHex(w.Size),
		toHex(w.Atmosphere),
		toHex(w.Hydro),
		toHex(w.Population),
		toHex(w.Government),
		toHex(w.Law),
		toHex(w.Tech))
}

type Companion struct {
	BaseOrbit
	CompanionStar *Star
}

func (c *Companion) GetUWP() string {
	return fmt.Sprintf("%-9s", c.CompanionStar.Classification())
}

func toHex(n int) string {
	if n < 0 {
		return "0"
	}
	if n < 10 {
		return fmt.Sprintf("%d", n)
	}
	hexMap := map[int]string{
		10: "A", 11: "B", 12: "C", 13: "D", 14: "E", 15: "F",
	}
	if h, ok := hexMap[n]; ok {
		return h
	}
	return "F"
}